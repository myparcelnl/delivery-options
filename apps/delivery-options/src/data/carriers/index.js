export * from './fetchAllCarriers';
export * from './fetchCarrierData';
export * from './createCarrierData';
export * from './formatCarrierResponse';
export * from './dhlCarrierConfiguration';
export * from './dpdCarrierConfiguration';
export * from './bpostCarrierConfiguration';
export * from './postNlCarrierConfiguration';
export * from './carrierConfigurationFactory';
export * from './abstractCarrierConfiguration';
export * from './instaboxCarrierConfiguration';
export * from './dhlForYouCarrierConfiguration';
export * from './cheapCargoCarrierConfiguration';
export * from './dhlEuroplusCarrierConfiguration';
export * from './dhlParcelConnectCarrierConfiguration';
