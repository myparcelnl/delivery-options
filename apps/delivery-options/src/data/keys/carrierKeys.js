/** @type {MyParcel.CarrierName} */
export const POSTNL = 'postnl';

/** @type {MyParcel.CarrierName} */
export const DPD = 'dpd';

/** @type {MyParcel.CarrierName} */
export const BPOST = 'bpost';

/** @type {MyParcel.CarrierName} */
export const INSTABOX = 'instabox';

/** @type {MyParcel.CarrierName} */
export const CHEAP_CARGO = 'cheapcargo';

/** @type {MyParcel.CarrierName} */
export const DHL = 'dhl';

/** @type {MyParcel.CarrierName} */
export const DHL_FOR_YOU = 'dhlforyou';

/** @type {MyParcel.CarrierName} */
export const DHL_PARCEL_CONNECT = 'dhlparcelconnect';

/** @type {MyParcel.CarrierName} */
export const DHL_EUROPLUS = 'dhleuroplus';
