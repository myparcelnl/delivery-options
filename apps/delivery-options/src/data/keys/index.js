export * from './configKeys';
export * from './addressKeys';
export * from './carrierKeys';
export * from './stringsKeys';
export * from './countryCodes';
export * from './platformKeys';
export * from './settingsConsts';
export * from './packageTypeConfig';

export * as ADDRESS from './addressKeys';

export * as CARRIERS from './carrierKeys';

export * as CONFIG from './configKeys';

export * as CONSTS from './settingsConsts';

export * as STRINGS from './stringsKeys';

export * as PLATFORMS from './platformKeys';

