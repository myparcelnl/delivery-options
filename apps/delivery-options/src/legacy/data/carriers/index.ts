export * from './createCarrierData';
export * from './fetchAllCarriers';
export * from './fetchCarrierData';
export * from './formatCarrierResponse';
export * from './resolveCarrierName';
