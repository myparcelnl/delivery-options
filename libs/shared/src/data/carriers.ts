import {BPOST, DHL_EUROPLUS, DHL_FOR_YOU, DHL_PARCEL_CONNECT, DPD, POSTNL, UPS} from '../legacy';

export const CARRIERS_MYPARCELNL = [POSTNL, DHL_FOR_YOU, DHL_EUROPLUS, DHL_PARCEL_CONNECT, UPS, DPD] as const;

export const CARRIERS_SENDMYPARCEL = [POSTNL, BPOST] as const;
