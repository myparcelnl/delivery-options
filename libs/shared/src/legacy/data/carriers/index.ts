export * from './abstractCarrierConfiguration';
export * from './bpostCarrierConfiguration';
export * from './dhlCarrierConfiguration';
export * from './dhlEuroplusCarrierConfiguration';
export * from './dhlForYouCarrierConfiguration';
export * from './dhlParcelConnectCarrierConfiguration';
export * from './dpdCarrierConfiguration';
export * from './postNlCarrierConfiguration';
export * from './upsCarrierConfiguration';
