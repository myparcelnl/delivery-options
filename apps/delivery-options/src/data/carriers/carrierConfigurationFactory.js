import {CARRIERS} from '../../data';
import { BpostCarrierConfiguration } from './bpostCarrierConfiguration';
import { CheapCargoCarrierConfiguration } from './cheapCargoCarrierConfiguration';
import { DhlCarrierConfiguration } from './dhlCarrierConfiguration';
import { DhlEuroplusCarrierConfiguration } from './dhlEuroplusCarrierConfiguration';
import { DhlForYouCarrierConfiguration } from './dhlForYouCarrierConfiguration';
import { DhlParcelConnectCarrierConfiguration } from './dhlParcelConnectCarrierConfiguration';
import { DpdCarrierConfiguration } from './dpdCarrierConfiguration';
import { InstaboxCarrierConfiguration } from './instaboxCarrierConfiguration';
import { PostNlCarrierConfiguration } from './postNlCarrierConfiguration';
import{ memoize } from 'lodash-unified';

const carrierConfiguration = memoize((carrierName, platform) => {
  switch (carrierName) {
    case CARRIERS.BPOST:
      return new BpostCarrierConfiguration(platform);
    case CARRIERS.CHEAP_CARGO:
      return new CheapCargoCarrierConfiguration(platform);
    case CARRIERS.DHL:
      return new DhlCarrierConfiguration(platform);
    case CARRIERS.DHL_FOR_YOU:
      return new DhlForYouCarrierConfiguration(platform);
    case CARRIERS.DHL_PARCEL_CONNECT:
      return new DhlParcelConnectCarrierConfiguration(platform);
    case CARRIERS.DHL_EUROPLUS:
      return new DhlEuroplusCarrierConfiguration(platform);
    case CARRIERS.DPD:
      return new DpdCarrierConfiguration(platform);
    case CARRIERS.POSTNL:
      return new PostNlCarrierConfiguration(platform);
    case CARRIERS.INSTABOX:
      return new InstaboxCarrierConfiguration(platform);
    default:
      throw new Error(`No configuration found for carrier ${carrierName}`);
  }
}, (carrierName, platform) => `${carrierName}_${platform}`);

export class CarrierConfigurationFactory {
  /**
   * @param {MyParcel.CarrierName} carrierName
   * @param {MyParcel.Platform} platform
   *
   * @returns {AbstractCarrierConfiguration}
   */
  static create(carrierName, platform) {
    return carrierConfiguration(carrierName, platform);
  }
}
