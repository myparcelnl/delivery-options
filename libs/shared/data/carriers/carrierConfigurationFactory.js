import * as CARRIERS from '@/data/keys/carrierKeys';
import { BpostCarrierConfiguration } from '@/data/carriers/bpostCarrierConfiguration';
import { DhlCarrierConfiguration } from '@/data/carriers/dhlCarrierConfiguration';
import { DhlEuroplusCarrierConfiguration } from '@/data/carriers/dhlEuroplusCarrierConfiguration';
import { DhlForYouCarrierConfiguration } from '@/data/carriers/dhlForYouCarrierConfiguration';
import { DhlParcelConnectCarrierConfiguration } from '@/data/carriers/dhlParcelConnectCarrierConfiguration';
import { DpdCarrierConfiguration } from '@/data/carriers/dpdCarrierConfiguration';
import { PostNlCarrierConfiguration } from '@/data/carriers/postNlCarrierConfiguration';
import memoize from 'lodash-es/memoize';

const carrierConfiguration = memoize((carrierName, platform) => {
  switch (carrierName) {
    case CARRIERS.BPOST:
      return new BpostCarrierConfiguration(platform);
    case CARRIERS.DHL:
      return new DhlCarrierConfiguration(platform);
    case CARRIERS.DHL_FOR_YOU:
      return new DhlForYouCarrierConfiguration(platform);
    case CARRIERS.DHL_PARCEL_CONNECT:
      return new DhlParcelConnectCarrierConfiguration(platform);
    case CARRIERS.DHL_EUROPLUS:
      return new DhlEuroplusCarrierConfiguration(platform);
    case CARRIERS.DPD:
      return new DpdCarrierConfiguration(platform);
    case CARRIERS.POSTNL:
      return new PostNlCarrierConfiguration(platform);
    default:
      throw new Error(`No configuration found for carrier ${carrierName}`);
  }
}, (carrierName, platform) => `${carrierName}_${platform}`);

export class CarrierConfigurationFactory {
  /**
   * @type {MyParcel.Platform|null}
   */
  platform = null;

  /**
   * @param {MyParcel.CarrierName} carrierName
   * @param {MyParcel.Platform|null} platform
   *
   * @returns {AbstractCarrierConfiguration}
   */
  static create(carrierName, platform = null) {
    this.platform = platform ?? this.platform;

    return carrierConfiguration(carrierName, this.platform);
  }
}
