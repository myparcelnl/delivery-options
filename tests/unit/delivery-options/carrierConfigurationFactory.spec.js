import * as CARRIERS from '@/data/keys/carrierKeys';
import { CarrierConfigurationFactory } from '@/data/carriers/carrierConfigurationFactory';
import { MYPARCEL } from '@/data/keys/platformKeys';

describe('CarrierConfigurationFactory', () => {
  it.each(Object.values(CARRIERS))('should create a configuration for %s', (carrier) => {
    const config = CarrierConfigurationFactory.create(carrier, MYPARCEL);

    expect({
      name: config.getName(),
      features: config.getFeatures(),
      platformFeatures: config.getPlatformFeatures(),
      countriesForDelivery: config.getCountriesForDelivery(),
      countriesForPickup: config.getCountriesForPickup(),
    }).toMatchSnapshot();
  });

  it('throws error when requesting nonexistent carrier', () => {
    expect(() => CarrierConfigurationFactory.create('nonexistent-carrier')).toThrowError();
  });

  it('remembers the platform', () => {
    CarrierConfigurationFactory.create(CARRIERS.POSTNL, MYPARCEL);

    expect(() => CarrierConfigurationFactory.create(CARRIERS.POSTNL)).not.toThrow();
  });
});
