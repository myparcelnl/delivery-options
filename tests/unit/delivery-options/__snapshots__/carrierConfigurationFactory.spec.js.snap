// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarrierConfigurationFactory should create a configuration for bpost 1`] = `
Object {
  "countriesForDelivery": Array [
    "BE",
    "NL",
  ],
  "countriesForPickup": Array [
    "BE",
    "NL",
  ],
  "features": Object {
    "belgie": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
      Array [
        "allowSaturdayDelivery",
        "fridayCutoffTime",
      ],
      Array [
        "allowSignature",
        "priceSignature",
      ],
      Array [
        "allowShowDeliveryDate",
      ],
      Array [
        "deliveryDaysWindow",
      ],
      Array [
        "dropOffDays",
      ],
      Array [
        "dropOffDelay",
      ],
      Array [
        "cutoffTime",
      ],
    ],
  },
  "name": "bpost",
  "platformFeatures": Array [],
}
`;

exports[`CarrierConfigurationFactory should create a configuration for dhl 1`] = `
Object {
  "countriesForDelivery": Array [
    "AT",
    "BG",
    "HR",
    "CZ",
    "DK",
    "EE",
    "FI",
    "FR",
    "DE",
    "GR",
    "HU",
    "IE",
    "IT",
    "LV",
    "LT",
    "PL",
    "PT",
    "RO",
    "SK",
    "SI",
    "ES",
    "SE",
  ],
  "countriesForPickup": Array [
    "AT",
    "BE",
    "DE",
    "DK",
    "FI",
    "FR",
    "GB",
    "PT",
  ],
  "features": Object {
    "myparcel": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
      Array [
        "allowShowDeliveryDate",
      ],
    ],
  },
  "name": "dhl",
  "platformFeatures": Array [
    Array [
      "allowDeliveryOptions",
      "priceStandardDelivery",
    ],
    Array [
      "allowPickupLocations",
      "pricePickup",
    ],
    Array [
      "allowShowDeliveryDate",
    ],
  ],
}
`;

exports[`CarrierConfigurationFactory should create a configuration for dhleuroplus 1`] = `
Object {
  "countriesForDelivery": Array [
    "BE",
    "BG",
    "DK",
    "DE",
    "EE",
    "FI",
    "FR",
    "GR",
    "HU",
    "IE",
    "IT",
    "HR",
    "LV",
    "LT",
    "LU",
    "NL",
    "AT",
    "PL",
    "PT",
    "RO",
    "SI",
    "SK",
    "ES",
    "CZ",
    "SE",
    "GB",
  ],
  "countriesForPickup": Array [],
  "features": Object {
    "myparcel": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
    ],
  },
  "name": "dhleuroplus",
  "platformFeatures": Array [
    Array [
      "allowDeliveryOptions",
      "priceStandardDelivery",
    ],
  ],
}
`;

exports[`CarrierConfigurationFactory should create a configuration for dhlforyou 1`] = `
Object {
  "countriesForDelivery": Array [
    "NL",
    "BE",
  ],
  "countriesForPickup": Array [
    "NL",
  ],
  "features": Object {
    "myparcel": Array [
      Array [
        "cutoffTime",
      ],
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "deliveryDaysWindow",
      ],
      Array [
        "dropOffDays",
      ],
      Array [
        "dropOffDelay",
      ],
      Array [
        "priceSameDayDelivery",
        "allowSameDayDelivery",
        "cutoffTimeSameDay",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
      Array [
        "priceOnlyRecipient",
        "allowOnlyRecipient",
      ],
      Array [
        "allowSignature",
        "priceSignature",
      ],
    ],
  },
  "name": "dhlforyou",
  "platformFeatures": Array [
    Array [
      "cutoffTime",
    ],
    Array [
      "allowDeliveryOptions",
      "priceStandardDelivery",
    ],
    Array [
      "deliveryDaysWindow",
    ],
    Array [
      "dropOffDays",
    ],
    Array [
      "dropOffDelay",
    ],
    Array [
      "priceSameDayDelivery",
      "allowSameDayDelivery",
      "cutoffTimeSameDay",
    ],
    Array [
      "allowPickupLocations",
      "pricePickup",
    ],
    Array [
      "priceOnlyRecipient",
      "allowOnlyRecipient",
    ],
    Array [
      "allowSignature",
      "priceSignature",
    ],
  ],
}
`;

exports[`CarrierConfigurationFactory should create a configuration for dhlparcelconnect 1`] = `
Object {
  "countriesForDelivery": Array [
    "BG",
    "DE",
    "EE",
    "FI",
    "GR",
    "HU",
    "IE",
    "IT",
    "HR",
    "LV",
    "LT",
    "LU",
    "AT",
    "PL",
    "PT",
    "RO",
    "SI",
    "SK",
    "ES",
    "CZ",
  ],
  "countriesForPickup": Array [
    "BG",
    "DK",
    "DE",
    "EE",
    "FI",
    "FR",
    "GR",
    "HU",
    "IE",
    "IT",
    "HR",
    "LV",
    "LT",
    "LU",
    "AT",
    "PL",
    "PT",
    "RO",
    "SI",
    "SK",
    "ES",
    "CZ",
    "SE",
    "GB",
  ],
  "features": Object {
    "myparcel": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
    ],
  },
  "name": "dhlparcelconnect",
  "platformFeatures": Array [
    Array [
      "allowDeliveryOptions",
      "priceStandardDelivery",
    ],
    Array [
      "allowPickupLocations",
      "pricePickup",
    ],
  ],
}
`;

exports[`CarrierConfigurationFactory should create a configuration for dpd 1`] = `
Object {
  "countriesForDelivery": Array [
    "AT",
    "BE",
    "BG",
    "CZ",
    "DK",
    "EE",
    "FI",
    "FR",
    "DE",
    "GR",
    "HU",
    "IE",
    "IT",
    "LV",
    "LI",
    "LT",
    "LU",
    "NL",
    "PL",
    "PT",
    "RO",
    "SK",
    "SI",
    "ES",
    "SE",
  ],
  "countriesForPickup": Array [
    "AT",
    "BE",
    "DK",
    "FI",
    "FR",
    "DE",
    "NL",
    "PT",
    "GB",
  ],
  "features": Object {
    "belgie": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
      Array [
        "dropOffDays",
      ],
      Array [
        "dropOffDelay",
      ],
      Array [
        "cutoffTime",
      ],
    ],
  },
  "name": "dpd",
  "platformFeatures": Array [],
}
`;

exports[`CarrierConfigurationFactory should create a configuration for postnl 1`] = `
Object {
  "countriesForDelivery": Array [
    "NL",
    "BE",
  ],
  "countriesForPickup": Array [
    "BE",
    "NL",
  ],
  "features": Object {
    "belgie": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "priceOnlyRecipient",
        "allowOnlyRecipient",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
      Array [
        "allowSignature",
        "priceSignature",
      ],
      Array [
        "allowShowDeliveryDate",
      ],
      Array [
        "deliveryDaysWindow",
      ],
      Array [
        "dropOffDays",
      ],
      Array [
        "dropOffDelay",
      ],
      Array [
        "cutoffTime",
      ],
    ],
    "myparcel": Array [
      Array [
        "allowDeliveryOptions",
        "priceStandardDelivery",
      ],
      Array [
        "allowEveningDelivery",
        "priceEveningDelivery",
      ],
      Array [
        "allowMorningDelivery",
        "priceMorningDelivery",
      ],
      Array [
        "priceOnlyRecipient",
        "allowOnlyRecipient",
      ],
      Array [
        "allowPackageTypeDigitalStamp",
      ],
      Array [
        "allowPackageTypeMailbox",
      ],
      Array [
        "allowMondayDelivery",
        "saturdayCutoffTime",
      ],
      Array [
        "allowPickupLocations",
        "pricePickup",
      ],
      Array [
        "allowSignature",
        "priceSignature",
      ],
      Array [
        "allowShowDeliveryDate",
      ],
      Array [
        "deliveryDaysWindow",
      ],
      Array [
        "dropOffDays",
      ],
      Array [
        "dropOffDelay",
      ],
      Array [
        "cutoffTime",
      ],
    ],
  },
  "name": "postnl",
  "platformFeatures": Array [
    Array [
      "allowDeliveryOptions",
      "priceStandardDelivery",
    ],
    Array [
      "allowEveningDelivery",
      "priceEveningDelivery",
    ],
    Array [
      "allowMorningDelivery",
      "priceMorningDelivery",
    ],
    Array [
      "priceOnlyRecipient",
      "allowOnlyRecipient",
    ],
    Array [
      "allowPackageTypeDigitalStamp",
    ],
    Array [
      "allowPackageTypeMailbox",
    ],
    Array [
      "allowMondayDelivery",
      "saturdayCutoffTime",
    ],
    Array [
      "allowPickupLocations",
      "pricePickup",
    ],
    Array [
      "allowSignature",
      "priceSignature",
    ],
    Array [
      "allowShowDeliveryDate",
    ],
    Array [
      "deliveryDaysWindow",
    ],
    Array [
      "dropOffDays",
    ],
    Array [
      "dropOffDelay",
    ],
    Array [
      "cutoffTime",
    ],
  ],
}
`;
